crates/zkvm/zkvm/src/prover.rs

```diff
@@ -1,50 +1,36 @@
-use std::fmt;
-
-/// Prover config of the ZkVm Host
-#[derive(Debug, Clone, Copy)]
-pub struct ProverOptions {
-    pub enable_compression: bool,
-    pub use_mock_prover: bool,
-    pub stark_to_snark_conversion: bool,
-    pub use_cached_keys: bool,
-}
+use crate::{host::ZkVmHost, input::ZkVmInputBuilder, proof::Proof, ProofType};
 
-// Compact representation of the prover options
-// Can be used to identify the saved proofs
-impl fmt::Display for ProverOptions {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let mut has_flags = false;
+pub trait ZkVmProver {
+    type Input;
+    type Output;
 
-        if self.enable_compression {
-            write!(f, "c")?;
-            has_flags = true;
-        }
+    fn proof_type() -> ProofType;
 
-        if self.use_mock_prover {
-            write!(f, "m")?;
-            has_flags = true;
-        }
+    /// Prepares the input for the zkVM.
+    fn prepare_input<'a, B>(input: &'a Self::Input) -> anyhow::Result<B::Input>
+    where
+        B: ZkVmInputBuilder<'a>;
 
-        if self.stark_to_snark_conversion {
-            write!(f, "s")?;
-            has_flags = true;
-        }
+    /// Processes the proof to produce the final output.
+    fn process_output<H>(proof: &Proof, host: &H) -> anyhow::Result<Self::Output>
+    where
+        H: ZkVmHost;
 
-        if has_flags {
-            write!(f, "_")?;
-        }
+    /// Proves the computation using any zkVM host.
+    fn prove<'a, H>(input: &'a Self::Input, host: &H) -> anyhow::Result<(Proof, Self::Output)>
```

Comment:
Yeah things like these should all use granularly typed errors like from `thiserror`.